### PR TITLE
Snake splicing

### PR DESCRIPTION
**Pull request by danking**

### CODE REVIEW

**Comment by danking** [comment]
This should be merged after #1751.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 1]
```diff
@@ -2519,37 +2519,142 @@ object FunctionRegistry {
...
+    .. code-block:: text
+        :emphasize-lines: 2
+
+        let a = [0, 2, 4, 6, 8, 10] in a[-3:-1]
+        result: [6, 8]
+
+    A handy way to understand the behavior of negative indicies is to recall
+    this rule:
+
+    .. code-block:: text
```
Comment (jigold) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: You need an empty line between the directive and the content.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 2]
```diff
@@ -2519,37 +2519,142 @@ object FunctionRegistry {
...
+        :emphasize-lines: 2
+
+        let a = "abcdef" in a[-3:-1]
+        result: "de"
+
+    A handy way to understand the behavior of negative indicies is to recall
+    this rule, which holds for any positive integers i and j:
+
+    .. code-block:: text
+        s[-i:-j] == s[s.length - i, s.length - j]
```
Comment (jigold) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Empty line between directive and content.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 3]
```diff
@@ -2575,37 +2575,68 @@ object FunctionRegistry {
         result: [0, 2, 4, 6]
     """
   )(arrayHr(TTHr), arrayHr(TTHr))
-  registerMethod("[*:]", (a: IndexedSeq[Any], i: Int) => a.slice(i, a.length),
+  registerMethod("[*:]", (a: IndexedSeq[Any], i: Int) => a.slice(if (i < 0) a.length + i else i, a.length),
     """
-    Returns a slice of the array from the i*th* element (0-indexed) to the end.
+    Returns a slice of the array from the i*th* element (0-indexed) to the
+    end. Negative indices are interpreted as offsets from the end of the string.
```
Comment (jigold) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: This is the indexed seq case. "Negative indices are interpreted as offsets from the last element in the array."

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 4]
```diff
@@ -2575,37 +2575,68 @@ object FunctionRegistry {
...
     """, "i" -> "Starting index of the slice."
   )(arrayHr(TTHr), intHr, arrayHr(TTHr))
-  registerMethod("[:*]", (a: IndexedSeq[Any], i: Int) => a.slice(0, i),
+  registerMethod("[:*]", (a: IndexedSeq[Any], i: Int) => a.slice(0, if (i < 0) a.length + i else i),
     """
-    Returns a slice of the array from the first element until the j*th* element (0-indexed).
+
+    Returns a slice of the array from the first element until the j*th* element
+    (0-indexed). Negative indices are interpreted as offsets from the end
+    of the string.
```
Comment (jigold) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: Same issue as above. Array and not String.

File: src/main/scala/is/hail/expr/FunctionRegistry.scala [hunk 5]
```diff
@@ -2575,37 +2575,68 @@ object FunctionRegistry {
...
+        result: [0, 2]
     """, "j" -> "End index of the slice (not included in result)."
   )(arrayHr(TTHr), intHr, arrayHr(TTHr))
-  registerMethod("[*:*]", (a: IndexedSeq[Any], i: Int, j: Int) => a.slice(i, j),
+  registerMethod("[*:*]", (a: IndexedSeq[Any], i: Int, j: Int) => a.slice(if (i < 0) a.length + i else i, if (j < 0) a.length + j else j),
     """
-    Returns a slice of the array from the i*th* element until the j*th* element (both 0-indexed).
+    Returns a slice of the array from the i*th* element until the j*th* element
+    (both 0-indexed). Negative indices are interpreted as offsets from the end
+    of the string.
```
Comment (jigold) [inline src/main/scala/is/hail/expr/FunctionRegistry.scala]: String => Array

**Comment by danking** [comment]
Good catch @jigold .

**Comment by danking** [comment]
@jigold back to you :)

**Pull request merged**